### PR TITLE
Feat/AN-4111 implement filter selection

### PR DESCRIPTION
**Pull request by MGJamJam**
Issue: https://bitmovin.atlassian.net/browse/AN-4111

Work: 
- Implements the filter Selection with Filter Validation and error message.
- Corrects wrong Tooltip text for OrderBy Selection. 
- Renames `MyDataSourceOptions` to `BitmovinDataSourceOptions`
image
image
image

### CODE REVIEW

File: bitmovin-analytics-datasource/src/types/queryFilter.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  { value: QUERY_FILTER_OPERATORS.CONTAINS, label: QUERY_FILTER_OPERATORS.CONTAINS },
+  { value: QUERY_FILTER_OPERATORS.NOTCONTAINS, label: QUERY_FILTER_OPERATORS.NOTCONTAINS },
+  { value: QUERY_FILTER_OPERATORS.IN, label: QUERY_FILTER_OPERATORS.IN },
+];
+
+export type QueryFilter = {
+  name: QueryAdAttribute | QueryAttribute;
+  operator: QueryFilterOperator;
+  value: QueryFilterValue;
+};
```
Comment (sabe25) [inline bitmovin-analytics-datasource/src/types/queryFilter.ts:35]: According to our api there is another property the filter can have. It's the `field` property where a reference to different data column can be set
https://github.com/bitmovin-engineering/analytics-resource-service/blob/b5aae2eb4d09275f75e80e8ace301979a711c49c/src/main/java/com/bitmovin/api/analytics/rest/dtos/request/QueryFilter.kt#L14-L15

Though i don't know if it is necessary to include this here.

**Review by sabe25 — APPROVED** [review]

File: bitmovin-analytics-datasource/src/types/queryFilter.ts (hunk 1 above)
Comment (MGJamJam) [reply bitmovin-analytics-datasource/src/types/queryFilter.ts:35]: We didn't include it in the old plugin, so for the first updated version I won't include it. But I will create a ticket to investigate if we should put it in. 👍

File: bitmovin-analytics-datasource/src/types/queryFilter.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+export type QueryFilterValue = boolean | number | string | string[] | null;
```
Comment (Svarozic) [inline bitmovin-analytics-datasource/src/types/queryFilter.ts:37]: `undefined` double check if that would not be a problem

File: bitmovin-analytics-datasource/src/utils/filterUtils.ts [hunk 3]
```diff
@@ -0,0 +1,186 @@
+import { QUERY_AD_ATTRIBUTES, QueryAdAttribute } from '../types/queryAdAttributes';
+import { QUERY_FILTER_OPERATORS, QueryFilterOperator, QueryFilterValue } from '../types/queryFilter';
+import { QUERY_ATTRIBUTES, QueryAttribute } from '../types/queryAttributes';
+
+export const isNullFilter = (filterAttribute: QueryAttribute | QueryAdAttribute): boolean => {
```
Comment (Svarozic) [inline bitmovin-analytics-datasource/src/utils/filterUtils.ts:5]:
```suggestion
const isNullFilter = (filterAttribute: QueryAttribute | QueryAdAttribute): boolean => {
```

File: bitmovin-analytics-datasource/src/types.ts [hunk 4]
```diff
@@ -15,19 +19,32 @@ export interface BitmovinAnalyticsDataQuery extends DataQuery {
...
  * These are options configured for each DataSource instance
  */
-export interface MyDataSourceOptions extends DataSourceJsonData {
+export interface BitmovinDataSourceOptions extends DataSourceJsonData {
   apiKey: string;
   tenantOrgId?: string;
   adAnalytics?: boolean;
 }
 
+export type BitmovinAnalyticsRequestQuery = {
```
Comment (Svarozic) [inline bitmovin-analytics-datasource/src/types.ts:36]: move it

File: bitmovin-analytics-datasource/src/utils/filterUtils.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Svarozic) [reply bitmovin-analytics-datasource/src/utils/filterUtils.ts:5]: use `isEmpty` from lodash

File: bitmovin-analytics-datasource/src/components/FilterRow.tsx [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+  selectedOperator: SelectableValue<QueryFilterOperator>;
+  rawFilterValue: string;
+  convertedFilterValue: QueryFilterValue;
+  parsingValueError: string;
+};
+
+export function FilterRow(props: Props) {
+  const [filters, setFilters] = useState<Filter[]>([]);
+
+  const mapFilterAttributesToSelectableValue = (): Array<SelectableValue<QueryAttribute | QueryAdAttribute>> => {
```
Comment (Svarozic) [inline bitmovin-analytics-datasource/src/components/FilterRow.tsx:28]: extract out as utility
```suggestion
  const mapFilterAttributesToSelectableValue = (filters: Filter[]): Array<SelectableValue<QueryAttribute | QueryAdAttribute>> => {
```

File: bitmovin-analytics-datasource/src/components/FilterRow.tsx [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+    const selectedAttributes = filters.map((filter) => filter.selectedAttribute);
+    if (props.isAdAnalytics) {
+      return difference(SELECTABLE_QUERY_AD_ATTRIBUTES, selectedAttributes);
+    } else {
+      return difference(SELECTABLE_QUERY_ATTRIBUTES, selectedAttributes);
+    }
+  };
+
+  const mapFiltersToQueryFilters = (filters: Filter[]): QueryFilter[] => {
```
Comment (Svarozic) [inline bitmovin-analytics-datasource/src/components/FilterRow.tsx:37]: extract out of component

File: bitmovin-analytics-datasource/src/components/FilterRow.tsx [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+
+      const newFilter = { ...filter, convertedFilterValue: convertedValue, parsingValueError: '' } as Filter;
+
+      const newFilters = [...filters];
+      newFilters.splice(index, 1, newFilter);
+
+      setFilters(newFilters);
+
+      props.onChange(mapFiltersToQueryFilters(newFilters));
+    } catch (e: any) {
```
Comment (Svarozic) [inline bitmovin-analytics-datasource/src/components/FilterRow.tsx:79]: like `new Error('Oh no!')` type

```suggestion
    } catch (e: Error) {
```

File: bitmovin-analytics-datasource/src/components/FilterRow.tsx [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+          selectableFilterAttributes={mapFilterAttributesToSelectableValue()}
+          onAttributeChange={(newValue: SelectableValue<QueryAdAttribute | QueryAttribute>) =>
+            onAttributesChange(index, newValue)
+          }
+          onOperatorChange={(newValue: SelectableValue<QueryFilterOperator>) => onOperatorsChange(index, newValue)}
+          onValueChange={(newValue: string) => onValuesChange(index, newValue)}
+          onDelete={() => deleteFilterInput(index)}
+          addFilterDisabled={isEmpty(filter.selectedAttribute) || isEmpty(filter.selectedOperator)}
+          onAddFilter={() => onAddFilter(index)}
+          parsingValueError={filter.parsingValueError === '' ? undefined : filter.parsingValueError}
```
Comment (Svarozic) [inline bitmovin-analytics-datasource/src/components/FilterRow.tsx:154]: `isEmpty()` , or `isBlank`

File: bitmovin-analytics-datasource/src/components/FilterInput.tsx [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+type Props = {
+  readonly isAdAnalytics: boolean;
+  readonly selectableFilterAttributes: Array<SelectableValue<QueryAttribute | QueryAdAttribute>>;
+  readonly onAttributeChange: (newValue: SelectableValue<QueryAdAttribute | QueryAttribute>) => void;
+  readonly onOperatorChange: (newValue: SelectableValue<QueryFilterOperator>) => void;
+  readonly onValueChange: (newValue: string) => void;
+  readonly onDelete: () => void;
+  readonly addFilterDisabled: boolean;
+  readonly onAddFilter: () => void;
+  readonly parsingValueError?: string;
```
Comment (Svarozic) [inline bitmovin-analytics-datasource/src/components/FilterInput.tsx:18]:
```suggestion
  readonly parsingValueError: string | undefined;
```

**Review by Svarozic — APPROVED** [review]

**Pull request merged**